### PR TITLE
crimson/osd: bring OpRegistry-based OpSequencer to handle out-of-order ClientRequests 

### PR DESCRIPTION
**Pull request by rzarzynski**
This is a severely reworked revision of PR #42661.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mattbenjamin** [comment]
@rzarzynski just randomly, why is any of this called a "registry?"

**Comment by rzarzynski** [comment]
@mattbenjamin: b/c it reuses the already available list-of-inflight-requests  (_OpRegistry_) instead of implementing something similar on its own.

**Comment by mattbenjamin** [comment]
> 
> 
> @mattbenjamin: b/c it reuses the already available list-of-inflight-requests (_OpRegistry_) instead of implementing something similar on its own.

I got that, but why is /that/ called a registry?  seems like a...list

**Comment by rzarzynski** [comment]
@mattbenjamin: ah, I see. I believe the key is in future extensions. Currently, it boils down into a list but more features (i.e. querying) will be added to form a counterpart of `OpTracker`.

**Comment by mattbenjamin** [comment]
> 
> 
> @mattbenjamin: ah, I see. I believe the key is in future extensions. Currently, it boils down into a list but more features (i.e. querying) will be added to form a counterpart of `OpTracker`.

wouldn't setting it up in an indexed data structure have excessive cost?

**Comment by rzarzynski** [comment]
@mattbenjamin: if you mean `OperationRegistryT::registries` then it shouldn't as it's a statically indexed (by an operation's `type`) `std::array` of `boost::intrusive::list`. If you mean collecting more data about particular op, they will be stored locally (op's members) to favor low recording overhead over the dumping performance.

**Comment by tchaikov** [comment]
i will finish the review tomorrow.

**Comment by athanatos** [comment]
@mattbenjamin It's a registry because every op is registered there.  *shrug* seems like a reasonable name to me.  I don't think "registry" inherently connotes a complex or expensive indexing scheme.

File: src/crimson/osd/osd_operation_sequencer.h [hunk 1]
```diff
@@ -26,71 +25,139 @@ namespace crimson::osd {
...
+
+  static bool is_unfinished(const ClientRequest& this_op) {
+    // TODO: kill the tombstone; reuse op status tracking.
+    return !this_op.finished;
+  }
+
+  std::uint64_t get_prev_id(const ClientRequest& this_op,
+                            const OSDOperationRegistry& registry) {
+    // an alternative to iterating till the registy's beginning could be
+    // holding a poihter to next(last_completed).
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation_sequencer.h:38]: s/poihter/pointer/

File: src/crimson/osd/osd_operation_sequencer.h [hunk 2]
```diff
@@ -26,71 +25,139 @@ namespace crimson::osd {
...
+    for (auto it = OSDOperationRegistry::op_list::s_iterator_to(this_op);
+         it != std::begin(registry.get_registry<type_idx>());
+         --it) {
+      // we're iterating over the operation registry of all ClientRequests.
+      // this list aggrates every single instance in the system, and thus
+      // we need to skip operations coming from different client's session
+      // or targeting different PGs.
+      // as this is supposed to happen on cold paths only, the overhead is
+      // a thing we can live with.
+      auto* maybe_prev_op = std::addressof(dynamic_cast<const ClientRequest&>(*it));
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation_sequencer.h:49]: i'd avoid using RTTI if possible.

File: src/crimson/osd/osd_operation_sequencer.h [hunk 3]
```diff
@@ -26,71 +25,139 @@ namespace crimson::osd {
...
+      if (maybe_prev_op->same_session_and_pg(this_op)) {
+        if (is_unfinished(*maybe_prev_op)) {
+          return maybe_prev_op->get_id();
+        } else {
+          // an early exited one
+          continue;
+        }
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation_sequencer.h:56]: i'd drop this branch. as we are skipping nothing with the `continue` clause.

File: src/crimson/osd/osd_operation_sequencer.h [hunk 4]
```diff
@@ -26,71 +25,139 @@ namespace crimson::osd {
...
-    assert(op.get_id() > last_completed);
-    last_completed = op.get_id();
+  void finish_op_out_of_order(ClientRequest& op,
+                              const OSDOperationRegistry& registry) {
+    ::crimson::get_logger(ceph_subsys_osd).debug(
+      "OpSequencer::{}: op={}, last_seen={}, last_unblocked={}, last_completed={}",
+      __func__, op.get_id(), last_seen_id, last_unblocked_id, last_completed_id);
+    // fix the last_unblocked_id. otherwise we wouldn't be able to leave
+    // the wait loop in start_op() as any concrete value of last_unblocked_id
+    // can wake at most one op and above we lowered this number to 0.
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation_sequencer.h:136]: i don't quite follow this line

> any concrete value of can wake at most one op and above we lowered this number to 0.

could you rephrase this sentence?

File: src/crimson/osd/osd_operation_sequencer.h [hunk 5]
```diff
@@ -99,18 +166,18 @@ class OpSequencer {
...
-  // |////|.....|.......| <--- last_issued
+  // |////|.....|.......| <--- last_seen
   //      ^     ^       ^
   //      |     |       \- prev_op
   //      |      \--- last_unblocked
   //      last_completed
   //
   // the id of last op which is issued
-  uint64_t last_issued = 0;
+  std::uint64_t last_seen_id = 0;
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation_sequencer.h:176]: hmm, i still think `last_issued` is a better name, but don't feel strong about this...

File: src/crimson/osd/osd_operation_sequencer.h [hunk 6]
```diff
@@ -25,68 +25,139 @@ namespace crimson::osd {
...
+        } else {
+          // an early exited one
+          continue;
+        }
+      }
+    }
+    // as all ops which haven't called `abort()` nor one of the finish()
+    // variants must be on the list, the only way we can reach here is
+    // the situation when the prev is the last completed op and it's aleady
+    // dead (== delisted).
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation_sequencer.h:62]: nit, 

> the prev op of this session targeting the same PG as this_op must has been completed and hence already has been removed from the list, that's the only way we got here

**Review by tchaikov — COMMENTED** [review]
IMHO, the last commit deserves some explanation in the commit message.

File: src/crimson/osd/osd_operation_sequencer.h (hunk 4 above)
Comment (rzarzynski) [reply src/crimson/osd/osd_operation_sequencer.h:136]: A blocked operation waits on its predecessor. We can also say an op can wake its successor but only if there is any (the op isn't the very last one) – that's why "_at most one_".

However, this order is violated by `finish_op_out_of_order()`. By removing an op-in-the-middle we're also taking out the perspective-of-being-waken from its successor ("_lowered this number to 0_").

Will try rephrase.

File: src/crimson/osd/osd_operation_sequencer.h (hunk 5 above)
Comment (rzarzynski) [reply src/crimson/osd/osd_operation_sequencer.h:176]: Ah, now I see: _last issued **by the client of OpSequncer**_. At first glance I was struggling to understand that. I think my view was more `OpSequencer`-centric (_last seen **by the OpSequncer**_).

Still, no strong feelings on the `last_seen` neither.

Comment (tchaikov) [reply src/crimson/osd/osd_operation_sequencer.h:176]: `last_seen` can be interpreted in multiple ways. if we consider this from the perspective of `OpSequencer`, what does "see" imply? if we can have a more accurate name, i'd be all for it.

**Review by tchaikov — APPROVED** [review]
apart from a nit, lgtm.

File: src/crimson/osd/osd_operation_sequencer.h (hunk 5 above)
Comment (rzarzynski) [reply src/crimson/osd/osd_operation_sequencer.h:176]: Hmm, basically it's the most recent op that `start_op()` has been called for. `last_started` maybe?

Nevertheless, I would be fine with reverting to `last_issued` too.

Comment (tchaikov) [reply src/crimson/osd/osd_operation_sequencer.h:176]: `last_stared` is much better.

Comment (rzarzynski) [reply src/crimson/osd/osd_operation_sequencer.h:176]: Switching!
